### DISCUSSION TITLE
Volume computation of pyramid elements

### DISCUSSION THREAD

**Discussion by ksHamburg** [pyvista/pyvista#6698] [Q&A] [answered]
### First check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the PyVista documentation, with the integrated search.
- [X] I already searched in Google "How to X in PyVista" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I will NOT make duplicate discussions or issues in the future.

### Commit to Help

- [X] I commit to help with one of those options.

### Sample Code What is the problem, question, or error?
Write a short description telling me what you are doing, what you expect to happen, and what is currently happening.

```python
Hi everyone,
I am trying to compute the volume of a mesh containing pyramid elements, since the pyvista implementation of computing the volume of a mesh is rather slow.
My approach is to divide all elements into tetrahedral elements and then compute the volume of these tetrahedral elements and sum everything up.
This approach works fine and fast for all elements except pyramids. For pyramids there are two options to divide a pyramid element into two tetrahedral elements.
For cases where the bottom face of the pyramid is not planar these two variants yield different volumina. 
Following my example code, I figured out, that pyvista is using the smallest volume of the two variants. 
However, when I generalize the volume computation of an entire mesh, I get different results to pyvistas volume computation.
So how does pyvista computes volumes of pyramids with non-planar base faces?
```

### Description

```python
import numpy as np
import pyvista as pv

pnts1 = np.array([[3.5      , 2.5588424, 3.5      ],
                  [3.4031935, 2.75     , 3.5      ],
                  [3.25     , 2.657165 , 3.75     ],
                  [3.262312 , 2.5      , 3.75     ],
                  [3.5      , 2.75     , 3.5      ]])
pyramid1 = pv.Pyramid([pnts1[0], pnts1[1], pnts1[2], pnts1[3], pnts1[4]])


pnts2 = np.array([[6.25    , 4.086113, 5.      ],
                  [6.29308 , 4.25    , 4.75    ],
                  [6.5     , 4.197463, 4.75    ],
                  [6.375809, 4.      , 5.      ],
                  [6.25    , 4.      , 5.      ]])
pyramid2 = pv.Pyramid([pnts2[0], pnts2[1], pnts2[2], pnts2[3], pnts2[4]])

def tet_vol(nds,ele):
    ms = np.hstack( ( nds[ele][:,1,:]-nds[ele][:,0,:],
                      nds[ele][:,2,:]-nds[ele][:,0,:],
                      nds[ele][:,3,:]-nds[ele][:,0,:],))
    spats = 1/6*abs(np.linalg.det(ms.reshape(ms.shape[0],3,3)))
    return sum(spats)

#
def pyramid_vol1(pyramid_mesh):
    ele = pyramid_mesh.cells.reshape( (pyramid_mesh.n_cells,6) )[:,1:]
    ele = np.vstack( (  ele[:,[0,2,4,1]],
                        ele[:,[0,2,4,3]], ))
    return tet_vol(pyramid_mesh.points,ele)
def pyramid_vol2(pyramid_mesh):
    ele = pyramid_mesh.cells.reshape( (pyramid_mesh.n_cells,6) )[:,1:]
    ele = np.vstack( (  ele[:,[0,1,3,4]],
                        ele[:,[1,2,3,4]], ))
    return tet_vol(pyramid_mesh.points,ele)

print('pyvista volumes:')
print( 'pyramid1.volume = ',pyramid1.volume )
print( 'pyramid2.volume = ',pyramid2.volume )

print('method 1 pyramid volumes:')
print( pyramid_vol1(pyramid1) )
print( pyramid_vol1(pyramid2) )

print('method 2 pyramid volumes:')
print( pyramid_vol2(pyramid1) )
print( pyramid_vol2(pyramid2) )


plotter = pv.Plotter()
poly = pv.PolyData(pyramid1.points)
poly["My Labels"] = [f"Label {i}" for i in range(pyramid1.n_points)]
plotter.add_point_labels(poly, "My Labels", point_size=20, font_size=36)
plotter.add_mesh(pyramid1, show_edges=True,color='white')
plotter.show()


plotter = pv.Plotter()
poly = pv.PolyData(pyramid2.points)
poly["My Labels"] = [f"Label {i}" for i in range(pyramid2.n_points)]
plotter.add_point_labels(poly, "My Labels", point_size=20, font_size=36)
plotter.add_mesh(pyramid2, show_edges=True,color='white')
plotter.show()


```

### System Information

```shell
--------------------------------------------------------------------------------
  Date: Fri Oct 11 14:55:45 2024 CEST

                  OS : Linux
              CPU(s) : 8
             Machine : x86_64
        Architecture : 64bit
                 RAM : 46.8 GiB
         Environment : Python
         File system : ext4
          GPU Vendor : NVIDIA Corporation
        GPU Renderer : NVIDIA GeForce MX450/PCIe/SSE2
         GPU Version : 4.5.0 NVIDIA 535.183.01
    MathText Support : False

  Python 3.10.13 (main, Sep  5 2023, 06:03:44) [GCC 11.4.0]

             pyvista : 0.44.1
                 vtk : 9.3.1
               numpy : 1.24.4
          matplotlib : 3.5.1
              scooby : 0.7.1
               pooch : v1.7.0
              pillow : 9.5.0
             imageio : 2.28.0
           pyvistaqt : 0.10.0
               PyQt5 : 5.15.6
             IPython : 8.12.1
            colorcet : 3.1.0
          ipywidgets : 7.8.4
               scipy : 1.11.2
                tqdm : 4.65.0
              meshio : 5.3.4
          jupyterlab : 4.0.2
               trame : 3.6.5
        trame_client : 3.2.5
        trame_server : 3.1.2
           trame_vtk : 2.8.10
       trame_vuetify : 2.7.1
jupyter_server_proxy : 4.0.0
        nest_asyncio : 1.5.6
```

### Screenshots

![pyramid2](https://github.com/user-attachments/assets/1a752e51-70cd-47e4-a0be-26889ff55e3c)
![pyramid1](https://github.com/user-attachments/assets/7901aa58-584f-450e-a3a7-3a74ac2b4d9f)

**Comment by MatthewFlamm** [comment]
vtk and thus PyVista expect a pyramid cell to have a planar quadrilateral base.  See: https://vtk.org/doc/nightly/html/classvtkPyramid.html#details

> a 3D cell that represents a linear pyramid

>> [vtkPyramid](https://vtk.org/doc/nightly/html/classvtkPyramid.html) is a concrete implementation of [vtkCell](https://vtk.org/doc/nightly/html/classvtkCell.html) to represent a 3D pyramid. A pyramid consists of a rectangular base with four triangular faces. [vtkPyramid](https://vtk.org/doc/nightly/html/classvtkPyramid.html) uses the standard isoparametric shape functions for a linear pyramid. The pyramid is defined by the five points (0-4) where (0,1,2,3) is the base of the pyramid which, using the right hand rule, forms a quadrilaterial whose normal points in the direction of the pyramid apex at vertex #4. The parametric location of vertex #4 is [0, 0, 1].

**Reply by MatthewFlamm** [reply]
Do you get the expected result for a planar base pyramid?

**Reply by MatthewFlamm** [reply]
To answer your question about how vtk tetrahedralizes the cell, see:

https://github.com/Kitware/VTK/blob/e684407dc721af41c93dc922169a69c0db18110d/Common/DataModel/vtkPyramid.cxx#L785-L803

**Reply by ksHamburg** [reply]
Hey MatthewFlamm,
thanks for your answers. 
When the bottom face of the pyramid is a flat quadrilateral, I get the expected result.

I now understand where the problem is:
`// The base of the pyramid must be split into two triangles.  There are two`
`// ways to do this (across either diagonal).  Pick the shorter diagonal.`

This is how vtk decides which subdivision to use. Now, I can replicate the exact same volume computation as PyVista gives me.
However, I am wondering how to treat pyramids with non-planar bases, since they are a real thing... I cannot see any evidence, that vtk or PyVista expect planar quadrilateral bases. Also, in the code snippet you send me, there is no such statement. 

I actually got these non-planar pyramids by using some filters on my mesh-data ( `mesh.clip_scalar( 'someScalar', value=x, invert=False ).extract_cells_by_type(pv.CellType.PYRAMID).volume` ). So PyVista i.e. vtk produces such non-planar pyramids itself.
In fact I think one should use numerical integration in conjunction with the  standard isoparametric shape functions of the pyramid elements. This might seem a little bit over engineered but would be the correct way... don't you think?

**Reply by MatthewFlamm** [reply]
> where (0,1,2,3) is the base of the pyramid which, using the right hand rule, forms a quadrilaterial whose  normal points in the direction of the pyramid apex at vertex #4. The parametric location of vertex #4 is [0, 0, 1].

This part of the description makes it seem to me that some operations expect a planar base.  A non planar base doesn't have an easily defined normal direction.  My experience is that while vtk allows certain things to happen, it doesn't necessarily support all operations for those situations.  If you have a better methodology, it would be great to submit upstream to vtk and then all downstream packages would benefit.

**Comment by ksHamburg** [comment, marked as answer]
After some digging I found some interesting things. 
First, [Felippa](https://www.researchgate.net/publication/242021889_A_compendium_of_FEM_integration_formulas_for_symbolic_work) describes isoparametric pyramid elements and writes "One of the faces is a quadrilateral, called the base, which may be warped.". 
To that end it is only possible to compute the volume via tetrahedrals if the base is planar otherwise one has to compute the exact integral which reads:
\int_P 1 dx dy dz = \int_0^1 \int_0^1 \int_0^1 det(J) drdsdt 
with the jacobian
J = dN(r,s,t).T @ points

Luckily, this integral can be exactly evaluated via numerical integration with only one integration point. To that end, I extendend my code (see below). As you can see if you run the code, there is a substantial difference in the computed volumes of roughly 25%. In my opinion one should always use the numerical integration version. This should also be fast, since we only have to evaluate one support point weight pair.

However, I am not familiar on how to contribute to the VTK library. What do you mean by submit to upstream?  Do you mean I should file an issue or report a bug?

```python
import numpy as np 
import sympy as sy
import pyvista as pv

pnts1 = np.array([[3.5      , 2.5588424, 3.5      ],
                  [3.4031935, 2.75     , 3.5      ],
                  [3.25     , 2.657165 , 3.75     ],
                  [3.262312 , 2.5      , 3.75     ],
                  [3.5      , 2.75     , 3.5      ]])
pyramid1 = pv.Pyramid([pnts1[0], pnts1[1], pnts1[2], pnts1[3], pnts1[4]])


pnts2 = np.array([[6.25    , 4.086113, 5.      ],
                  [6.29308 , 4.25    , 4.75    ],
                  [6.5     , 4.197463, 4.75    ],
                  [6.375809, 4.      , 5.      ],
                  [6.25    , 4.      , 5.      ]])
pyramid2 = pv.Pyramid([pnts2[0], pnts2[1], pnts2[2], pnts2[3], pnts2[4]])

def tet_vol(nds,ele):
    ms = np.hstack( ( nds[ele][:,1,:]-nds[ele][:,0,:],
                      nds[ele][:,2,:]-nds[ele][:,0,:],
                      nds[ele][:,3,:]-nds[ele][:,0,:],))
    spats = 1/6*abs(np.linalg.det(ms.reshape(ms.shape[0],3,3)))
    return sum(spats)
def pyramid_vol1(pyramid_mesh):
    ele = pyramid_mesh.cells.reshape( (pyramid_mesh.n_cells,6) )[:,1:]
    ele = np.vstack( (  ele[:,[0,2,4,1]],
                        ele[:,[0,2,4,3]], ))
    return tet_vol(pyramid_mesh.points,ele)
def pyramid_vol2(pyramid_mesh):
    ele = pyramid_mesh.cells.reshape( (pyramid_mesh.n_cells,6) )[:,1:]
    ele = np.vstack( (  ele[:,[0,1,3,4]],
                        ele[:,[1,2,3,4]], ))
    return tet_vol(pyramid_mesh.points,ele)
def pyramid_vol(nds):
    r,s,t = 1/2,1/2,np.sqrt(3)/3+1 # support point of adopted
    w     = 1 # weight of adopted numerical integration scheme 
    # derivatives of the shape function
    dN    = np.array([ [-(1-s)*(1-t),-(1-r)*(1-t),-(1-r)*(1-s)],
                       [ (1-s)*(1-t),-r*(1-t),-r*(1-s)],
                       [ s*(1-t),r*(1-t),-r*s],
                       [ -s*(1-t),(1-r)*(1-t),-(1-r)*s],
                       [ 0,0,1]
                    ])
    return np.linalg.det( dN.T @ nds ) # determinant of the jacobian
def pyramid_volume_analytical(nds):
    r,s,t = sy.symbols('r,s,t')
    N = sy.Matrix([ (1-r)*(1-s)*(1-t),
                       r *(1-s)*(1-t),
                       r * s   *(1-t),
                    (1-r)* s   *(1-t),
                    t
                  ]) 
    dN   = sy.Matrix([ [-(1-s)*(1-t),-(1-r)*(1-t),-(1-r)*(1-s)],
                       [ (1-s)*(1-t),    -r*(1-t),    -r*(1-s)],
                       [     s*(1-t),     r*(1-t),    -r*s    ],
                       [    -s*(1-t), (1-r)*(1-t),-(1-r)*s    ],
                       [           0,           0,           1]])
    Jac  = sy.simplify( dN.T @ nds )
    detJ = sy.simplify( Jac.det() ) 
    vol  = sy.integrate(detJ,(r,0,1),(s,0,1),(t,0,1))
    ## derivation of support points and weights
    # general pyramid coordinates
    x0,y0,z0 = sy.symbols('x0,y0,z0')
    x1,y1,z1 = sy.symbols('x1,y1,z1')
    x2,y2,z2 = sy.symbols('x2,y2,z2')
    x3,y3,z3 = sy.symbols('x3,y3,z3')
    x4,y4,z4 = sy.symbols('x4,y4,z4')
    nds = sy.Matrix([[x0,y0,z0],
                     [x1,y1,z1],
                     [x2,y2,z2],
                     [x3,y3,z3],
                     [x4,y4,z4],])
    Jac         = dN.T @ nds
    detJ        = Jac.det()
    vol_generic = sy.integrate(detJ,(r,0,1),(s,0,1),(t,0,1))
    # detJ is linear in r and linear in s -> 1 point integration rule
    print('degree of detJ in r : ',sy.degree(detJ,r), ' -> r_i,w_i = 1/2,1' )
    print('degree of detJ in s : ',sy.degree(detJ,s), ' -> r_i,w_i = 1/2,1')
    
    print('degree of detJ in t : ',sy.degree(detJ,t))
    print('get factors of detJ at support points r_i,s_i')
    print( sy.factor_list(detJ.subs({r:sy.S(1)/sy.S(2),s:sy.S(1)/sy.S(2)}))[1][0] )
    print('quadratic dependence of detJ w.r.t. t is: (1-t)**2')
    print('this enables us to solve: int_0^1 a(t-1)**2 dt = a(t-1)**2')
    a = sy.symbols('a')
    print( 't_i,w_i : ',sy.solve(sy.integrate( a*(1-t)**2,(t,0,1)) - (1-t)**2*a,t)[1],',0' )
    return vol

print('pyvista volumes:')
print( 'pyramid1.volume = ',pyramid1.volume )
print( 'pyramid2.volume = ',pyramid2.volume )

print('method 1 pyramid volumes:')
print( pyramid_vol1(pyramid1) )
print( pyramid_vol1(pyramid2) )

print('method 2 pyramid volumes:')
print( pyramid_vol2(pyramid1) )
print( pyramid_vol2(pyramid2) )


print('analytical volumes via num. int.:')
print( pyramid_vol(pnts1) )
print( pyramid_vol(pnts2) )

print('pyramid 1: volume difference in %: ',100-pyramid1.volume/pyramid_vol(pnts1)*100 )
print('pyramid 2: volume difference in %: ',100-pyramid2.volume/pyramid_vol(pnts2)*100 )

pyramid_volume_analytical(pnts1)
```

**Reply by MatthewFlamm** [reply]
This is the finite element view of the world, and I'm not so sure it is the same for all other types of cells.  It might be, but I haven't thought about it deeply.  It seems ill-posed from a finite volume perspective, for which I am more familiar.

In any case, my point about upstream is that PyVista directly uses vtk for this calculation, see https://github.com/pyvista/pyvista/blob/6624cdedf0416d7dfb4c300328a43f2f98f07c3a/pyvista/core/filters/data_set.py#L2403-L2410
I do not think we should build in special cases in PyVista for a bug (or wanted enhancement) in vtk, rather that it should be fixed or implemented in vtk.  I can give you a pointer that the volume is calculated in vtk for a generic 3D cell [here](https://github.com/Kitware/VTK/blob/b7f436cdc98673c1a1d6f17e69cfd8646da19726/Filters/Verdict/vtkCellSizeFilter.cxx#L307-L312).  For these cells, it tetrahedralizes them, then calculates the volume of each tetrahedra, see [IntegrateGeneral3DCell](https://github.com/Kitware/VTK/blob/b7f436cdc98673c1a1d6f17e69cfd8646da19726/Filters/Verdict/vtkCellSizeFilter.cxx#L683).  There are a bunch of special cased cells where the volume is determined through another method, for example a Voxel see [case statement](https://github.com/Kitware/VTK/blob/b7f436cdc98673c1a1d6f17e69cfd8646da19726/Filters/Verdict/vtkCellSizeFilter.cxx#L234-L241) and [IntegrateVoxel](https://github.com/Kitware/VTK/blob/b7f436cdc98673c1a1d6f17e69cfd8646da19726/Filters/Verdict/vtkCellSizeFilter.cxx#L660).  If you wanted to special case a Pyramid, I assume it would follow similarly, but with your implementation.  I would recommend that you first open an issue on the [vtk gitlab site](https://gitlab.kitware.com/vtk/vtk).

BTW the Pyramid cell seems to get less love, as I had recently found the centroid calculation was very wrong.  see https://gitlab.kitware.com/vtk/vtk/-/issues/18972.

**Reply by ksHamburg** [reply]
Yes you are right, it is truly a Finite Element view I am exploring here. However, all these considerations are based on the shape functions of the element which I found following your very first link to the vtk source code.
Anyhow, I see how I can get in touch with the vtk-guys. The problem is, that I am not a C++ programmer. I will see how it goes. In the meantime I simply use my suggestion for my code. 

And yes you are right. Pyramid elements play a minor role in computational physics since they also have bad convergence properties in the context of Finite Elements unless one uses some "rabbit" functions. Due to that, I was amazed that the pyvista (or vtk under the hood) produced a pyramid element polluted mesh after a applying the "clip_scalar" filter...
